File: Esp32S3/main/protocol/protocol.c
#include "protocol.h"
#include <string.h>
#include <stdio.h>
#include "esp_log.h"
#include "esp_timer.h"
#include "json_lite.h"
#include "nav_frame.h"
#include "config.h"
#include "battery.h"   /* 【M7】电量上报（vbat/batt） */
#include "lcd_driver.h"
#include "render_nav.h"
#include "map_image.h"     /* 【M3.1】IMG_BEGIN/CHUNK/END 的接收接口 */

static const char *TAG = "proto";
static uint32_t s_err = 0;

/* 应用层链路状态：只看“收到什么报文”（依据唯一，不依赖 TCP 连接是否抖动） */
#define LINK_TIMEOUT_US (5LL * 1000000LL)       /* 5 秒无任何报文 => App 不在了（异常断开兜底） */
static int64_t s_t_msg = 0;                     /* 最近收到任意报文的时刻 */
static int64_t s_t_nav = 0;                     /* 最近收到 NAV_FRAME 的时刻 */

/* 纯报文驱动：只要“App 的报文还在”就算已连接（画面内容完全由报文决定） */
int protocol_link_stage(void)
{
    int64_t now = esp_timer_get_time();
    if (s_t_msg == 0 || (now - s_t_msg) > LINK_TIMEOUT_US) return LINK_STAGE_WAIT_APP;
    return LINK_STAGE_CONNECTED;
}

void protocol_link_reset(void)
{
    s_t_msg = 0;
    s_t_nav = 0;
}

uint32_t protocol_err_count(void) { return s_err; }
void     protocol_err_reset(void) { s_err = 0; }

/* 回 DEV_STATUS（协议 §4.1）：配置 + 版本 + 错误计数
 * 【M1】追加 7 项叠加层可读性参数，App 保存后用 GET_CONFIG 回读校验一致性。
 * buf 由 240 提到 384：新增字段约 110 字符，避免 snprintf 截断（-Werror=format-truncation）。 */
static void send_dev_status(proto_send_fn send, void *ctx)
{
    const espnav_config_t *c = config_get();
    char buf[512];   /* 【M7】追加 img_on 等字段后防截断（原 384 是 M1 时的余量）*/
    snprintf(buf, sizeof(buf),
             "{\"msg_type\":\"DEV_STATUS\",\"payload\":{"
             "\"lcd_brightness\":%u,\"dash_speed\":%u,\"anim_enable\":%s,"
             "\"popup_timeout\":%u,"
             "\"scrim_on\":%s,\"scrim_compass\":%u,\"scrim_text\":%u,"
             "\"scrim_route\":%u,\"scrim_clock\":%u,\"grid_bright\":%u,\"map_area\":%u,"
             "\"img_on\":%s,\"screen_flip_y\":%s,"
             "\"vbat\":%d,\"batt\":%d,"
             "\"firmware_ver\":\"%s\",\"err\":%lu}}\n",
             (unsigned)c->lcd_brightness, (unsigned)c->dash_speed,
             c->anim_enable ? "true" : "false", (unsigned)c->popup_timeout,
             c->scrim_on ? "true" : "false",
             (unsigned)c->scrim_compass, (unsigned)c->scrim_text,
             (unsigned)c->scrim_route, (unsigned)c->scrim_clock,
             (unsigned)c->grid_bright, (unsigned)c->map_area,
             c->img_on ? "true" : "false", c->screen_flip_y ? "true" : "false",
             (int)battery_mv(), battery_pct(),
             c->firmware_ver, (unsigned long)s_err);
    if (send) send(buf, ctx);
    ESP_LOGI(TAG, "TX DEV_STATUS bright=%u dash=%u anim=%d popup=%u scrim=%d/%u,%u,%u,%u grid=%u area=%u ver=%s err=%lu",
             (unsigned)c->lcd_brightness, (unsigned)c->dash_speed, (int)c->anim_enable,
             (unsigned)c->popup_timeout, (int)c->scrim_on,
             (unsigned)c->scrim_compass, (unsigned)c->scrim_text,
             (unsigned)c->scrim_route, (unsigned)c->scrim_clock,
             (unsigned)c->grid_bright, (unsigned)c->map_area,
             c->firmware_ver, (unsigned long)s_err);
}

/* SET_CONFIG（协议 §3.2）：逐项应用，立即生效 */
static void apply_set_config(const char *line)
{
    int v = 0;
    bool on = false;

    if (jl_get_int(line, "lcd_brightness", -1, &v) && v >= 0) {
        config_set_brightness((uint8_t)v);
        const lcd_driver_ops_t *ops = lcd_driver_ops();
        if (ops && ops->backlight) ops->backlight((uint8_t)v);
        ESP_LOGI(TAG, "apply lcd_brightness=%d", v);
    }
    if (jl_get_int(line, "dash_speed", -1, &v) && v >= 0) {
        config_set_dash_speed((uint8_t)v);          /* 渲染 tick 每帧读 config，下一帧即生效 */
        ESP_LOGI(TAG, "apply dash_speed=%d", v);
    }
    if (jl_get_bool(line, "anim_enable", &on)) {
        config_set_anim_enable(on);                 /* false = 静态虚线（不推进相位） */
        ESP_LOGI(TAG, "apply anim_enable=%d", (int)on);
    }
    if (jl_get_int(line, "popup_timeout", -1, &v) && v >= 0) {
        config_set_popup_timeout((uint8_t)v);       /* 弹窗层 M7 使用 */
        ESP_LOGI(TAG, "apply popup_timeout=%d", v);
    }
    /* ---- M1：叠加层可读性（App 设置项下发；渲染每帧读 config，下一帧即生效）---- */
    if (jl_get_bool(line, "scrim_on", &on)) {
        config_set_scrim_on(on);
        ESP_LOGI(TAG, "apply scrim_on=%d", (int)on);
    }
    if (jl_get_int(line, "scrim_compass", -1, &v) && v >= 0) {
        config_set_scrim_compass((uint8_t)v);
        ESP_LOGI(TAG, "apply scrim_compass=%d", v);
    }
    if (jl_get_int(line, "scrim_text", -1, &v) && v >= 0) {
        config_set_scrim_text((uint8_t)v);
        ESP_LOGI(TAG, "apply scrim_text=%d", v);
    }
    if (jl_get_int(line, "scrim_route", -1, &v) && v >= 0) {
        config_set_scrim_route((uint8_t)v);
        ESP_LOGI(TAG, "apply scrim_route=%d", v);
    }
    if (jl_get_int(line, "scrim_clock", -1, &v) && v >= 0) {
        config_set_scrim_clock((uint8_t)v);
        ESP_LOGI(TAG, "apply scrim_clock=%d", v);
    }
    if (jl_get_int(line, "grid_bright", -1, &v) && v >= 0) {
        config_set_grid_bright((uint8_t)v);
        ESP_LOGI(TAG, "apply grid_bright=%d", v);
    }
    if (jl_get_int(line, "map_area", -1, &v) && v >= 0) {
        config_set_map_area((uint8_t)v);
        ESP_LOGI(TAG, "apply map_area=%d", v);
    }
    /* ---- 【M2.4】ESP 屏颜色（0 = 用固件默认色；渲染端每帧读 config）---- */
    if (jl_get_int(line, "col_main", -1, &v) && v >= 0) {
        config_set_col_main((uint16_t)v);
        ESP_LOGI(TAG, "apply col_main=%d", v);
    }
    if (jl_get_int(line, "col_track", -1, &v) && v >= 0) {
        config_set_col_track((uint16_t)v);
        ESP_LOGI(TAG, "apply col_track=%d", v);
    }
    if (jl_get_int(line, "col_grid", -1, &v) && v >= 0) {
        config_set_col_grid((uint16_t)v);
        ESP_LOGI(TAG, "apply col_grid=%d", v);
    }
    if (jl_get_int(line, "col_road", -1, &v) && v >= 0) {
        config_set_col_road((uint16_t)v);
        ESP_LOGI(TAG, "apply col_road=%d", v);
    }
    if (jl_get_int(line, "col_car", -1, &v) && v >= 0) {
        config_set_col_car((uint16_t)v);
        ESP_LOGI(TAG, "apply col_car=%d", v);
    }
    if (jl_get_int(line, "col_hint", -1, &v) && v >= 0) {
        config_set_col_hint((uint16_t)v);
        ESP_LOGI(TAG, "apply col_hint=%d", v);
    }
    /* ---- 【M2.5】分光镜 HUD：整屏水平镜像（App 设置项，默认开）---- */
    if (jl_get_bool(line, "screen_flip", &on)) {
        config_set_screen_flip(on);
        lcd_ili9341_set_flip_x(on);              /* 直接作用于驱动，下次刷屏即生效 */
        ESP_LOGI(TAG, "apply screen_flip=%d", (int)on);
    }
    /* ---- 【M9】整屏垂直镜像：与水平翻转独立，可单独开、也可同时开 ---- */
    if (jl_get_bool(line, "screen_flip_y", &on)) {
        config_set_screen_flip_y(on);
        lcd_ili9341_set_flip_y(on);              /* 直接作用于驱动，下次刷屏即生效 */
        ESP_LOGI(TAG, "apply screen_flip_y=%d", (int)on);
    }
    /* ---- 【M7】APK 地图底图总开关：false = 丢弃底图回原始导航模式 ----
     * 只改配置；渲染侧（draw_frame，仅显示任务碰帧缓冲）看到 false 时自行 map_image_clear()，
     * 因此这里不跨线程调用渲染侧接口。 */
    if (jl_get_bool(line, "img_on", &on)) {
        config_set_img_on(on);
        ESP_LOGI(TAG, "apply img_on=%d (0=drop map image, back to template road)", (int)on);
    }
}

void protocol_handle(const char *line, int len, proto_send_fn send, void *ctx)
{
    char mt[20];

    if (!line || len <= 0) return;
    if (!jl_get_str(line, "msg_type", mt, sizeof(mt))) {
        s_err++;                                     /* 非法帧计数（协议 §4.1 err） */
        ESP_LOGW(TAG, "RX len=%d no msg_type -> err=%lu | %.60s", len, (unsigned long)s_err, line);
        return;
    }
    ESP_LOGI(TAG, "RX len=%d type=%s", len, mt);     /* 探针：每条收到的报文都留痕（NUL 已由 frame_parser 补齐） */

    s_t_msg = esp_timer_get_time();                  /* 任何合法报文都算“App 在”（BYE 除外，见上） */

    if (!strcmp(mt, "BYE")) {                        /* 主动断开：立即回“等待 App” */
        protocol_link_reset();
        ESP_LOGI(TAG, "RX BYE -> 立即回到“等待手机App连接”");
        return;
    }
    if (!strcmp(mt, "HELLO")) {                      /* 握手：App 上线声明 */
        const espnav_config_t *c = config_get();
        char hb[128];
        snprintf(hb, sizeof(hb),
                 "{\"msg_type\":\"HELLO_ACK\",\"payload\":{\"firmware_ver\":\"%s\"}}\n",
                 c->firmware_ver);
        if (send) send(hb, ctx);
        ESP_LOGI(TAG, "RX HELLO -> TX HELLO_ACK ver=%s", c->firmware_ver);
        return;
    }
    if (!strcmp(mt, "NAV_FRAME")) {                  /* 仅缓存，渲染由显示任务完成 */
        s_t_nav = s_t_msg;
        nav_frame_on_json_line(line, len);
        return;
    }
    if (!strcmp(mt, "CLOCK")) {                      /* 【M2.6】时间下发（ESP 无 RTC）：待机/导航画面都显示 */
        char cb[NAV_CLOCK_MAX + 1];
        if (jl_get_str(line, "clock", cb, sizeof(cb))) {
            render_nav_set_clock(cb);
            ESP_LOGI(TAG, "RX CLOCK %s", cb);
        }
        return;
    }
    if (!strcmp(mt, "PING")) {
        int ts = 0, v = 0;
        jl_get_int(line, "ts", 0, &v);
        ts = v;
        char buf[96];
        snprintf(buf, sizeof(buf), "{\"msg_type\":\"PONG\",\"payload\":{\"ts\":%d}}\n", ts);
        if (send) send(buf, ctx);
        ESP_LOGI(TAG, "TX PONG ts=%d", ts);
        return;
    }
    if (!strcmp(mt, "GET_CONFIG")) {                  /* §3.3 -> DEV_STATUS，并清零 err */
        send_dev_status(send, ctx);
        s_err = 0;
        return;
    }
    if (!strcmp(mt, "SET_CONFIG")) {
        apply_set_config(line);
        send_dev_status(send, ctx);                   /* 回状态作为“已生效”确认（协议未定义 ACK） */
        return;
    }
    if (!strcmp(mt, "CLEAR_SCREEN")) {                /* §3.4 黑屏待机 */
        render_nav_clear();
        ESP_LOGI(TAG, "CLEAR_SCREEN done");
        return;
    }
    if (!strcmp(mt, "POPUP_MSG") || !strcmp(mt, "POPUP_CLOSE")) {
        ESP_LOGW(TAG, "%s 尚未实现（M7 弹窗层），已忽略", mt);
        return;
    }
    /* ---- 【M3.1】地图截图通道：接收侧只累积 base64 字节 + 置标志，
     * 不碰 framebuffer（解码/贴图由显示任务在 draw_frame 里做，见 fault_log F5 线程约定）---- */
    if (!strcmp(mt, "IMG_BEGIN")) {
        int seq = 0, w = 0, h = 0, x = 0, y = 0, bytes = 0;
        jl_get_int(line, "seq", 0, &seq);
        jl_get_int(line, "w", 320, &w);
        jl_get_int(line, "h", 240, &h);
        jl_get_int(line, "x", 0, &x);
        jl_get_int(line, "y", 0, &y);
        jl_get_int(line, "bytes", 0, &bytes);
        map_image_begin(seq, w, h, x, y, bytes);
        return;
    }
    if (!strcmp(mt, "IMG_CHUNK")) {
        static char b64[MAP_IMG_B64_MAX];               /* static：不占 TCP 任务栈 */
        int seq = 0;
        jl_get_int(line, "seq", 0, &seq);
        if (jl_get_str(line, "d", b64, sizeof(b64))) map_image_chunk(seq, b64);
        else ESP_LOGW(TAG, "IMG_CHUNK 缺 d 字段");
        return;
    }
    if (!strcmp(mt, "IMG_END")) {
        int seq = 0;
        jl_get_int(line, "seq", 0, &seq);
        map_image_end(seq);
        return;
    }
    ESP_LOGD(TAG, "ignore unknown msg_type=%s（协议 §6.2）", mt);
}

File: mobileApp/android/app/src/main/java/com/espnav/app/protocol/Protocol.kt
package com.espnav.app.protocol

import org.json.JSONArray
import org.json.JSONObject

/**
 * 协议层（V1.10 / V1.14）：出站报文构建与入站报文解析。
 * 仅做 JSON 组装/提取，不涉及网络（网络见 net/EspNavClient.kt）。
 */

/** 路况模板（协议 §3.1.1；字段与固件 nav_road_t、HTML V2 readRoad 对齐） */
data class RoadSpec(
    val type: String,
    val dir: String? = null,
    val exits: List<String>? = null,
    val pts: List<Pair<Int, Int>>? = null,
    val half: Int? = null,
    val cx: Int? = null,
    val cy: Int? = null,
    val r: Int? = null
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("type", type)
        if (dir != null) put("dir", dir)
        if (exits != null) put("exits", JSONArray(exits))
        if (pts != null) put("pts", Json.pts(pts))   // 已 smart-cast，无需 !!
        if (half != null) put("half", half)
        if (cx != null) put("cx", cx)
        if (cy != null) put("cy", cy)
        if (r != null) put("r", r)
    }
}

/** 导航帧（协议 §3.1） */
data class NavFrame(
    val heading: Int = 0,
    val turnDist: Int = 0,
    val hint: String = "",
    val totalDist: Int = 0,
    val progressPct: Int = 0,
    val elapsedMin: Int = 0,
    val etaTime: String = "",
    val centerLine: List<Pair<Int, Int>> = emptyList(),
    val pastCenter: List<Pair<Int, Int>> = emptyList(),
    val routeCenter: List<Pair<Int, Int>> = emptyList(),
    val pos: Pair<Int, Int>? = null,
    val overview: List<Pair<Int, Int>> = emptyList(),
    val overviewDot: Pair<Int, Int>? = null,
    val road: RoadSpec? = null,
    /** 当前速度（km/h，供固件/串口日志显示） */
    val speedKmh: Int = 0,
    /** 当前路段名称（固件显示在屏幕第一行，超长自动滚动） */
    val roadName: String = "",
    /** 提示文本（路况/测速等；固件显示在“距离”下方的黄色行） */
    val notice: String = "",
    /** 【M1.2】时:分:秒（"HH:MM:SS"）—— ESP 无 RTC，时间必须由 App 每帧下发；
     *  固件显示在主视图右下角，空串则不显示。 */
    val clock: String = ""
) {
    fun payloadJson(): JSONObject = JSONObject().apply {
        put("heading", heading)
        put("turn_dist", turnDist)
        put("hint", hint)
        put("total_dist", totalDist)
        put("progress_pct", progressPct)
        put("elapsed_min", elapsedMin)
        put("eta_time", etaTime)
        put("centerLine", Json.pts(centerLine))
        put("pastCenter", Json.pts(pastCenter))
        put("routeCenter", Json.pts(routeCenter))
        pos?.let { put("pos", Json.pair(it)) }
        put("overview", Json.pts(overview))
        overviewDot?.let { put("overview_dot", Json.pair(it)) }
        road?.let { put("road", it.toJson()) }
        put("speed", speedKmh)
        if (roadName.isNotEmpty()) put("roadName", roadName)
        if (notice.isNotEmpty()) put("notice", notice)
        if (clock.isNotEmpty()) put("clock", clock)     /* 【M1.2】时间（主视图右下角） */
    }

    fun toJsonLine(): String =
        JSONObject().put("msg_type", "NAV_FRAME").put("payload", payloadJson()).toString()
}

/** JSON 小工具 */
object Json {
    fun pair(p: Pair<Int, Int>): JSONArray = JSONArray().put(p.first).put(p.second)

    fun pts(list: List<Pair<Int, Int>>): JSONArray =
        JSONArray().apply { list.forEach { put(pair(it)) } }

    fun obj(vararg kv: Pair<String, Any>): JSONObject =
        JSONObject().apply { kv.forEach { put(it.first, it.second) } }
}

/** 出站控制报文（协议 §3.2~§3.7） */
object OutMsg {
    fun navFrame(frame: NavFrame): String = frame.toJsonLine()

    /** 主动断开通知：ESP 收到后立即回到“等待手机App连接”（协议 V1.13） */
    fun bye(): String =
        Json.obj("msg_type" to "BYE", "payload" to JSONObject()).toString()

    /** 握手：App 上线声明（ESP 收到后回 HELLO_ACK，据此进入“已连接”阶段） */
    fun hello(): String =
        Json.obj("msg_type" to "HELLO", "payload" to JSONObject()).toString()

    fun ping(ts: Long): String =
        Json.obj("msg_type" to "PING", "payload" to Json.obj("ts" to ts)).toString()

    /** 【M2.6】时间下发：ESP 无 RTC，App 连接后每秒发一次；待机画面也据此显示时间 */
    fun clock(hhmmss: String): String =
        Json.obj("msg_type" to "CLOCK", "payload" to Json.obj("clock" to hhmmss)).toString()

    /* ---- 【M3.1】地图底图分块传输（协议 V1.14）----
     * IMG_BEGIN(seq,w,h,x,y,bytes) → N×IMG_CHUNK(seq, d=base64) → IMG_END(seq)。
     * 固件按 seq 丢旧保新；单帧 ≤2048B（每块原始 ≤1200 → base64 ≤1600 字符）。 */
    fun imgBegin(seq: Int, w: Int, h: Int, x: Int, y: Int, bytes: Int): String =
        Json.obj(
            "msg_type" to "IMG_BEGIN",
            "payload" to Json.obj(
                "seq" to seq, "w" to w, "h" to h, "x" to x, "y" to y,
                "fmt" to "jpg", "bytes" to bytes
            )
        ).toString()

    fun imgChunk(seq: Int, b64: String): String =
        Json.obj("msg_type" to "IMG_CHUNK", "payload" to Json.obj("seq" to seq, "d" to b64)).toString()

    fun imgEnd(seq: Int): String =
        Json.obj("msg_type" to "IMG_END", "payload" to Json.obj("seq" to seq)).toString()

    fun getConfig(): String =
        Json.obj("msg_type" to "GET_CONFIG", "payload" to JSONObject()).toString()

    fun clearScreen(): String =
        Json.obj("msg_type" to "CLEAR_SCREEN", "payload" to JSONObject()).toString()

    /** SET_CONFIG（协议 §3.2）：只下发非 null 的字段。
     *  【M1】新增 6 项 ESP 屏叠加层可读性参数（scrim_* / grid_bright），
     *  语义：透明度 0=全黑底衬、100=不铺底衬。 */
    fun setConfig(
        brightness: Int? = null,
        dashSpeed: Int? = null,
        animEnable: Boolean? = null,
        popupTimeout: Int? = null,
        scrimOn: Boolean? = null,
        scrimCompass: Int? = null,
        scrimText: Int? = null,
        scrimRoute: Int? = null,
        scrimClock: Int? = null,
        gridBright: Int? = null,
        screenFlip: Boolean? = null,
        /* 【M9】整屏垂直镜像（上下翻转） */
        screenFlipY: Boolean? = null,
        /* 【M2.4】ESP 屏颜色（RGB565 十进制；0 = 用固件默认色） */
        colMain: Int? = null,
        colTrack: Int? = null,
        colGrid: Int? = null,
        colRoad: Int? = null,
        colCar: Int? = null,
        colHint: Int? = null,
        /* 【M7】APK 地图底图总开关：false = 固件丢弃底图、回到原始导航模式 */
        imgOn: Boolean? = null
    ): String {
        val p = JSONObject()
        brightness?.let { p.put("lcd_brightness", it) }
        dashSpeed?.let { p.put("dash_speed", it) }
        animEnable?.let { p.put("anim_enable", it) }
        popupTimeout?.let { p.put("popup_timeout", it) }
        scrimOn?.let { p.put("scrim_on", it) }
        scrimCompass?.let { p.put("scrim_compass", it) }
        scrimText?.let { p.put("scrim_text", it) }
        scrimRoute?.let { p.put("scrim_route", it) }
        scrimClock?.let { p.put("scrim_clock", it) }
        gridBright?.let { p.put("grid_bright", it) }
        /* 【M2.5】分光镜 HUD：整屏水平镜像 */
        screenFlip?.let { p.put("screen_flip", it) }
        screenFlipY?.let { p.put("screen_flip_y", it) }
        colMain?.let { p.put("col_main", it) }
        colTrack?.let { p.put("col_track", it) }
        colGrid?.let { p.put("col_grid", it) }
        colRoad?.let { p.put("col_road", it) }
        colCar?.let { p.put("col_car", it) }
        colHint?.let { p.put("col_hint", it) }
        imgOn?.let { p.put("img_on", it) }
        return Json.obj("msg_type" to "SET_CONFIG", "payload" to p).toString()
    }
}

/** 入站报文解析（DEV_STATUS / PONG；未知类型保持原样） */
object InMsg {
    fun msgType(line: String): String =
        runCatching { JSONObject(line).optString("msg_type", "") }.getOrDefault("")

    fun describe(line: String): String = runCatching {
        val o = JSONObject(line)
        val type = o.optString("msg_type", "?")
        val p = o.optJSONObject("payload") ?: JSONObject()
        when (type) {
            "DEV_STATUS" ->
                "DEV_STATUS 亮度=${p.optInt("lcd_brightness")} 速度=${p.optInt("dash_speed")} " +
                    "动画=${p.optBoolean("anim_enable")} 超时=${p.optInt("popup_timeout")}s " +
                    "底衬=${p.optBoolean("scrim_on")}/${p.optInt("scrim_compass")}," +
                    "${p.optInt("scrim_text")},${p.optInt("scrim_route")},${p.optInt("scrim_clock")} " +
                    "网格=${p.optInt("grid_bright")} " +
                    "底图=${p.optBoolean("img_on")} " +
                    "电量=${p.optInt("batt")}%(${p.optInt("vbat")}mV) " +
                    "版本=${p.optString("firmware_ver")} err=${p.optInt("err")}"
            "PONG" -> "PONG ts=${p.optLong("ts")}"
            else -> line
        }
    }.getOrDefault(line)
}
